arangod/Aql/Functions.cpp

```diff
@@ -284,9 +284,50 @@ AqlValue numberValue(double value, bool nullify) {
   return AqlValue(AqlValueHintDouble(value));
 }
 
-inline AqlValue timeAqlValue(tp_sys_clock_ms const& tp) {
-  std::string formatted = format("%FT%TZ", floor<milliseconds>(tp));
-  return AqlValue(formatted);
+/// @brief optimized version of datetime stringification
+/// string format is hard-coded to YYYY-MM-DDTHH:MM:SS.XXXZ
+AqlValue timeAqlValue(tp_sys_clock_ms const& tp) {
+  char formatted[24];
+
+  year_month_day ymd{floor<days>(tp)};
+  auto day_time = make_time(tp - sys_days(ymd));
+
+  auto y = static_cast<int>(ymd.year());
+  formatted[0] = '0' + (y / 1000);
+  formatted[1] = '0' + ((y % 1000) / 100);
+  formatted[2] = '0' + ((y % 100) / 10);
+  formatted[3] = '0' + (y % 10);
+  formatted[4] = '-';
+  auto m = static_cast<unsigned>(ymd.month());
+  formatted[5] = '0' + (m / 10);
+  formatted[6] = '0' + (m % 10);
+  formatted[7] = '-';
+  auto d = static_cast<unsigned>(ymd.day());
+  formatted[8] = '0' + (d / 10);
+  formatted[9] = '0' + (d % 10);
+  formatted[10] = 'T';
+  auto h = day_time.hours().count();
+  formatted[11] = '0' + (h / 10);
+  formatted[12] = '0' + (h % 10);
+  formatted[13] = ':';
+  auto i = day_time.minutes().count();
+  formatted[14] = '0' + (i / 10);
+  formatted[15] = '0' + (i % 10);
+  formatted[16] = ':';
+  auto s = day_time.seconds().count();
+  formatted[17] = '0' + (s / 10);
+  formatted[18] = '0' + (s % 10);
+  formatted[19] = '.';
+  uint64_t millis = day_time.subseconds().count();
+  if (millis > 999) {
```

Comment:
why can this be larger then 999 ?

Reply:
This is existing code, just moved around somewhat. So the restriction was already in place.
I didn't want to change the behavior of existing code nor change guarantees in this PR.